### DISCUSSION TITLE
<Spinner>s when switching between certain pages only

### DISCUSSION THREAD

**Discussion by tyteen4a03** [vercel/next.js#14294] [Help] [answered]
I have a Page setup that looks like this:

/blog/[id].tsx
```
const BlogPost: NextPage<BlogPostProps> = ({ blogPost }) => (
    <>
        <Navbar />
        <Spinner />
        <div className="inner">
        {blogPost}
        </div>
    </>
);

... getStaicProps etc
```

When I switch from `/blog/1` to `/blog/2`, I want the `<div className="inner">` to not display and instead show the `<Spinner>` while the other page loads. Once the page finishes loading the `<Spinner>` should hide itself.

The problem with the solutions I see online is they are for global page transitions; it is not possible to only have the transition show for a defined page (I have other pages which does not need the spinner). How do I achieve the above effect?

**Comment by jamesmosier** [comment, marked as answer]
Could you do something similar to the [with-loading](https://github.com/vercel/next.js/blob/canary/examples/with-loading/pages/_app.js) example but do it on a per page basis? 

Maybe even creating a component like...

```jsx
export default function Layout({ children }) {
  const [showSpinner, setShowSpinner] = React.useState(false);
  const [loaded, setLoaded] = React.useState(false);

  React.useEffect(() => {
    const changeStart = url => {
      setShowSpinner(true);
    };
    const changeDone = () => {
      setShowSpinner(false);
      setLoaded(true);
    };

    Router.events.on('routeChangeStart', changeStart);
    Router.events.on('routeChangeComplete', changeDone);
    Router.events.on('routeChangeError', changeDone);

    return () => {
      Router.events.off('routeChangeStart', changeStart);
      Router.events.off('routeChangeComplete', changeDone);
      Router.events.off('routeChangeError', changeDone);
    };
  }, []);

  return (
    <>
      {showSpinner && <Spinner />}
      {loaded && <div className="inner">{children}</div>}
    </>
  );
}
```

Otherwise if you wanted it to be global, in your `_app.js` like the example, you could check what url you are on in the `routeChangeStart` event and then set some state accordingly.

**Reply by tyteen4a03** [reply]
I defintiely did not think of `off` - will give this a try!

**Comment by timmysmalls** [comment]
Hey @tyteen4a03 and @jamesmosier, I think I may have a similar problem, but I'm not sure. Here's what I'm facing.

I have a basic route: `[slug].js`, that will show categories of products. So for example `www.example.com/shoes`, or `www.example.com/shirts`. On these pages it will do a simple api call to retrieve fifteen products in that category. However, when I'm on `/shirts`, and I click a nextjs link to `/shoes`, the framework seems to recongnize that it's about to render the same page, just with a different parameter for `slug`. Instead of rerendering the entire page (which would be the expected behaviour, and the behaviour that I'm after), it keeps showing the products from the `shirts` page until the shoes are done loading.

Like I said, I'd expect the page to just rerender and start off with a spinner until the products are done loading, but it seems that rather than unmounting the page component and remounting it with new props, it just swaps out the old props for the new props, while leaving the page component mounted. This causes an unnatural delay when switching URLs. Preferably I'd like to only keep the components in `_app` mounted (navbar and footer), and just rerender everything else, to not have to worry about stale state lingering around.

Is this similar to what you were facing here?

**Reply by cdock1029** [reply]
@timmysmalls wrap what you want to be remounted inside a component then provide a `key` (probably the slug) which will force React to unmount and reload this component.

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key
